File: src/main/java/com/jcloisterzone/event/MeepleReturned.java
package com.jcloisterzone.event;

import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.figure.Meeple;
import com.jcloisterzone.game.ReturnMeepleSource;

public class MeepleReturned extends PlayEvent {

    private static final long serialVersionUID = 1L;

    private Meeple meeple;
    private FeaturePointer from;
    /** true if meeple is returned different way than scoring feature */
    private final boolean forced;
    private ReturnMeepleSource returnMeepleSource;

    public MeepleReturned(PlayEventMeta metadata, Meeple meeple, FeaturePointer from, boolean forced, ReturnMeepleSource returnMeepleSource) {
        super(metadata);
        this.meeple = meeple;
        this.from = from;
        this.forced = forced;
        this.returnMeepleSource = returnMeepleSource;
    }

    public Meeple getMeeple() {
        return meeple;
    }

    public FeaturePointer getFrom() {
        return from;
    }

    public boolean isForced() {
        return forced;
    }

    public ReturnMeepleSource getReturnMeepleSource() {
        return returnMeepleSource;
    }
}

File: src/main/ts/com/jcloisterzone/event/MeepleReturned.ts
import type { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import type { Meeple } from "../figure/Meeple.js";
import type { ReturnMeepleSource } from "../game/ReturnMeepleSource.js";
import { PlayEvent, type PlayEventMeta } from "./PlayEvent.js";

/** A meeple was returned to its owner's supply. */
export class MeepleReturned extends PlayEvent {
  static readonly simpleName = "MeepleReturned";

  constructor(
    metadata: PlayEventMeta,
    private readonly meeple: Meeple,
    private readonly from: FeaturePointer,
    /** true if returned by means other than scoring the feature */
    private readonly forced: boolean,
    private readonly returnMeepleSource: ReturnMeepleSource | null,
  ) {
    super(metadata);
  }

  getMeeple(): Meeple {
    return this.meeple;
  }
  getFrom(): FeaturePointer {
    return this.from;
  }
  isForced(): boolean {
    return this.forced;
  }
  getReturnMeepleSource(): ReturnMeepleSource | null {
    return this.returnMeepleSource;
  }
}
